README.rst
=======================
Guzzle Cache Subscriber
=======================

.. important::

    **This repo has not been updated for Guzzle 6 and only supports Guzzle 5.**

    See https://github.com/Kevinrob/guzzle-cache-middleware for a nice Guzzle 6
    compatible Cache middleware.

Provides a private transparent proxy cache for caching HTTP responses.

Here's a simple example of how it's used:

.. code-block:: php

    use GuzzleHttp\Client;
    use GuzzleHttp\Subscriber\Cache\CacheSubscriber;

    $client = new Client(['defaults' => ['debug' => true]]);

    // Use the helper method to attach a cache to the client.
    CacheSubscriber::attach($client);

    // Send the first request
    $a = $client->get('http://en.wikipedia.org/wiki/Main_Page');

    // Send the second request. This will find a cache hit which must be
    // validated. The validation request returns a 304, which yields the original
    // cached response.
    $b = $client->get('http://en.wikipedia.org/wiki/Main_Page');

Running the above code sample should output verbose cURL information that looks
something like this:

::

    > GET /wiki/Main_Page HTTP/1.1
    Host: en.wikipedia.org
    User-Agent: Guzzle/4.2.1 curl/7.37.0 PHP/5.5.13
    Via: 1.1 GuzzleCache/4.2.1

    < HTTP/1.1 200 OK
    < Server: Apache
    < X-Content-Type-Options: nosniff
    < Content-language: en
    < X-UA-Compatible: IE=Edge
    < Vary: Accept-Encoding,Cookie
    < Last-Modified: Thu, 21 Aug 2014 01:51:49 GMT
    < Content-Type: text/html; charset=UTF-8
    < X-Varnish: 2345493325, 1998949714 1994269567
    < Via: 1.1 varnish, 1.1 varnish
    < Transfer-Encoding: chunked
    < Date: Thu, 21 Aug 2014 02:34:12 GMT
    < Age: 2541
    < Connection: keep-alive
    < X-Cache: cp1055 hit (1), cp1068 frontend hit (25353)
    < Cache-Control: private, s-maxage=0, max-age=0, must-revalidate
    < Set-Cookie: GeoIP=US:Seattle:47.6062:-122.3321:v4; Path=/; Domain=.wikipedia.org
    <
    * Connection #0 to host en.wikipedia.org left intact
    * Re-using existing connection! (#0) with host en.wikipedia.org
    > GET /wiki/Main_Page HTTP/1.1
    Host: en.wikipedia.org
    User-Agent: Guzzle/4.2.1 curl/7.37.0 PHP/5.5.13
    Via: 1.1 GuzzleCache/4.2.1, 1.1 GuzzleCache/4.2.1
    If-Modified-Since: Thu, 21 Aug 2014 01:51:49 GMT

    < HTTP/1.1 304 Not Modified
    < Server: Apache
    < X-Content-Type-Options: nosniff
    < Content-language: en
    < X-UA-Compatible: IE=Edge
    < Vary: Accept-Encoding,Cookie
    < Last-Modified: Thu, 21 Aug 2014 01:51:49 GMT
    < Content-Type: text/html; charset=UTF-8
    < X-Varnish: 2345493325, 1998950450 1994269567
    < Via: 1.1 varnish, 1.1 varnish
    < Date: Thu, 21 Aug 2014 02:34:12 GMT
    < Age: 2541
    < Connection: keep-alive
    < X-Cache: cp1055 hit (1), cp1068 frontend hit (25360)
    < Cache-Control: private, s-maxage=0, max-age=0, must-revalidate
    < Set-Cookie: GeoIP=US:Seattle:47.6062:-122.3321:v4; Path=/; Domain=.wikipedia.org
    <
    * Connection #0 to host en.wikipedia.org left intact

Installing
----------

Add the following to your composer.json:

.. code-block:: javascript

    {
        "require": {
            "guzzlehttp/cache-subscriber": "0.2.*@dev"
        }
    }

or

.. code-block:: console

    $ composer require guzzlehttp/cache-subscriber

Creating a CacheSubscriber
--------------------------

The easiest way to create a CacheSubscriber is using the ``attach()`` helper
method of ``GuzzleHttp\Subscriber\Cache\CacheSubscriber``. This method accepts
a request or client object and attaches the necessary subscribers used to
perform cache lookups, validation requests, and automatic purging of resources.

The ``attach()`` method accepts the following options:

storage
    A ``GuzzleHttp\Subscriber\Cache\CacheStorageInterface`` object used to
    store cached responses. If no value is not provided, an in-memory array
    cache will be used.
validate
    A Boolean value that determines if cached response are ever validated
    against the origin server. This setting defaults to ``true`` but can be
    disabled by passing ``false``.
purge
    A Boolean value that determines if cached responses are purged when
    non-idempotent requests are sent to their URI. This setting defaults to
    ``true`` but can be disabled by passing ``false``.
can_cache
    An optional callable used to determine if a request can be cached. The
    callable accepts a ``GuzzleHttp\Message\RequestInterface`` and returns a
    Boolean value. If no value is provided, the default behavior is utilized.

.. warning::

    This is a WIP update for Guzzle 5+. It hasn't been tested and is in
    active development. Expect bugs and breaks.


CHANGELOG.md
# CHANGELOG

## 0.2.0 - 2019-09-16

* Improvement: Tests for expired items without must-revalidate #20
* Improvement: Support for including Vary headers in cache keys #21
* Improvement: Test for the can_cache option #23
* Bugfix: Not adding timezone to dates #24
* Improvement: Add $defaultTtl to CacheStorage constructor #25
* Bugfix: Error caches responses for without vary headers #26
* Bugfix: stale-if-header not being added to max-age #27
* Bugfix: max-age and freshness confusing zero and null #28
* Bugfix: Use date() method to fix missing GMT #29
* Improvement: Tests for stale-if-error behaviour #30
* Improvement: Delete cache entries on both 404 and 410 responses #33
* Refactoring: Minor ValidationSubscriber.php cleanup #35
* Refactoring: Minor ValidationSubscriber.php cleanup #35
* Improvement: Extend caching ttl considerations #56
* Improvement: Add purge method #57
* Improvement: Integration test for the calculation of the "resident_time" #60
* Improvement: Support for PHP 7.0 & 7.1 #73

## 0.1.0 - 2014-10-29

* Initial release.


src/ValidationSubscriber.php
<?php
namespace GuzzleHttp\Subscriber\Cache;

use GuzzleHttp\Event\CompleteEvent;
use GuzzleHttp\Event\RequestEvents;
use GuzzleHttp\Event\SubscriberInterface;
use GuzzleHttp\Exception\BadResponseException;
use GuzzleHttp\Message\RequestInterface;
use GuzzleHttp\Message\ResponseInterface;

/**
 * Validates cached responses as needed.
 *
 * @Link http://tools.ietf.org/html/rfc7234#section-4.3
 */
class ValidationSubscriber implements SubscriberInterface
{
    /** @var CacheStorageInterface Cache object storing cache data */
    private $storage;

    /** @var callable */
    private $canCache;

    /** @var array */
    private static $gone = [404 => true, 410 => true];

    /** @var array */
    private static $replaceHeaders = [
        'Date',
        'Expires',
        'Cache-Control',
        'ETag',
        'Last-Modified',
    ];

    /**
     * @param CacheStorageInterface $cache    Cache storage
     * @param callable              $canCache Callable used to determine if a
     *                                        request can be cached. Accepts a
     *                                        RequestInterface and returns a
     *                                        boolean value.
     */
    public function __construct(
        CacheStorageInterface $cache,
        callable $canCache
    ) {
        $this->storage = $cache;
        $this->canCache = $canCache;
    }

    public function getEvents()
    {
        return ['complete' => ['onComplete', RequestEvents::EARLY]];
    }

    public function onComplete(CompleteEvent $e)
    {
        $lookup = $e->getRequest()->getConfig()->get('cache_lookup');

        if ($lookup == 'HIT' &&
            $this->shouldvalidate($e->getRequest(), $e->getResponse())
        ) {
            $this->validate($e->getRequest(), $e->getResponse(), $e);
        }
    }

    private function validate(
        RequestInterface $request,
        ResponseInterface $response,
        CompleteEvent $event
    ) {
        try {
            $validate = $this->createRevalidationRequest($request, $response);
            $validated = $event->getClient()->send($validate);
        } catch (BadResponseException $e) {
            $this->handleBadResponse($e);
        }

        if ($validated->getStatusCode() == 200) {
            $this->handle200Response($request, $validated, $event);
        } elseif ($validated->getStatusCode() == 304) {
            $this->handle304Response($request, $response, $validated, $event);
        }
    }

    private function shouldValidate(
        RequestInterface $request,
        ResponseInterface $response
    ) {
        if ($request->getMethod() != 'GET'
            || $request->getConfig()->get('cache.disable')
        ) {
            return false;
        }

        $validate = Utils::getDirective($request, 'Pragma') === 'no-cache'
            || Utils::getDirective($response, 'Pragma') === 'no-cache'
            || Utils::getDirective($request, 'must-revalidate')
            || Utils::getDirective($response, 'must-revalidate')
            || Utils::getDirective($request, 'no-cache')
            || Utils::getDirective($response, 'no-cache')
            || Utils::getDirective($response, 'max-age') === '0'
            || Utils::getDirective($response, 's-maxage') === '0';

        // Use the strong ETag validator if available and the response contains
        // no Cache-Control directive
        if (!$validate
            && !$response->hasHeader('Cache-Control')
            && $response->hasHeader('ETag')
        ) {
            $validate = true;
        }

        return $validate;
    }

    /**
     * Handles a bad response when attempting to validate.
     *
     * If the resource no longer exists, then remove from the cache.
     *
     * @param BadResponseException $e Exception encountered
     *
     * @throws BadResponseException
     */
    private function handleBadResponse(BadResponseException $e)
    {
        if (isset(self::$gone[$e->getResponse()->getStatusCode()])) {
            $this->storage->delete($e->getRequest());
        }

        throw $e;
    }

    /**
     * Creates a request to use for revalidation.
     *
     * @param RequestInterface  $request  Request
     * @param ResponseInterface $response Response to validate
     *
     * @return RequestInterface returns a revalidation request
     */
    private function createRevalidationRequest(
        RequestInterface $request,
        ResponseInterface $response
    ) {
        $validate = clone $request;
        $validate->getConfig()->set('cache.disable', true);
        $validate->removeHeader('Pragma');
        $validate->removeHeader('Cache-Control');
        $responseDate = $response->getHeader('Last-Modified')
            ?: $response->getHeader('Date');
        $validate->setHeader('If-Modified-Since', $responseDate);

        if ($etag = $response->getHeader('ETag')) {
            $validate->setHeader('If-None-Match', $etag);
        }

        return $validate;
    }

    private function handle200Response(
        RequestInterface $request,
        ResponseInterface $validateResponse,
        CompleteEvent $event
    ) {
        // Store the 200 response in the cache if possible
        if (Utils::canCacheResponse($validateResponse)) {
            $this->storage->cache($request, $validateResponse);
        }

        $event->intercept($validateResponse);
    }

    private function handle304Response(
        RequestInterface $request,
        ResponseInterface $response,
        ResponseInterface $validated,
        CompleteEvent $event
    ) {
        // Make sure that this response has the same ETag
        if ($validated->getHeader('ETag') !== $response->getHeader('ETag')) {
            // Revalidation failed, so remove from cache and retry.
            $this->storage->delete($request);
            $event->intercept($event->getClient()->send($request));

            return;
        }

        // Replace cached headers with any of these headers from the
        // origin server that might be more up to date
        $modified = false;
        foreach (self::$replaceHeaders as $name) {
            if ($validated->hasHeader($name)
                && $validated->getHeader($name) != $response->getHeader($name)
            ) {
                $modified = true;
                $response->setHeader($name, $validated->getHeader($name));
            }
        }

        // Store the updated response in cache
        if ($modified) {
            $this->storage->cache($request, $response);
        }
    }
}
